### ISSUE TITLE
BN+ReLU cause "RuntimeError: MALFORMED INPUT: bad dtype in CompareSelect" error in fp16, traced module

### ISSUE THREAD

**Issue by GrzegorzKarchNV** [pytorch/pytorch#61382]
## 🐛 Bug

Using a combination of BatchNorm1d and relu in a traced module causes the "bad dtype in CompareSelect" when executed in fp16 mode. 

## To Reproduce

1. run the PyTorch code:

```python
import torch.nn.functional as F
import torch
import torch.nn as nn

class JasperBlock(nn.Module):
    def __init__(self):
        super(JasperBlock, self).__init__()
        self.bn = nn.BatchNorm1d(3, eps=1e-3, momentum=0.1)
        
    def forward(self, x):
        out = self.bn(x)
        out = F.relu(out)        
        return out

if __name__ == "__main__":

    model = JasperBlock()
    model.cuda()
    model.eval()
    model = model.half()

    x = torch.randn((3,3))*10
    x = x.cuda().half()
    out = model(x)
    model_jit = torch.jit.trace(model, (x,))
    out_jit = model_jit(x) 
```

Output:
```
Traceback (most recent call last):
  File "inference.py", line 417, in <module>
    main()
  File "inference.py", line 357, in main
    log_probs = model(feats)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
RuntimeError: MALFORMED INPUT: bad dtype in CompareSelect
```

## Expected behavior

I expect to successfully run the code without any error.

## Environment

PyTorch version: 1.9.0a0+c3d40fd
Is debug build: False
CUDA used to build PyTorch: 11.3
OS: Ubuntu 20.04.2 LTS (x86_64)
Docker container: nvcr.io/nvidia/pytorch:21.06-py3

cc @albanD @mruberry @jbschlosser @gmagogsfm

**Comment by GrzegorzKarchNV** [comment]
the last working version was PyTorch Version 1.9.0a0+2ecb2c7 from nvcr.io/nvidia/pytorch:21.05-py3

**Comment by gmagogsfm** [comment]
@ZolotukhinM I think you know tensor expression the best, could you help take a look?

**Comment by ZolotukhinM** [comment]
Yes, I'll take a look, thanks for the report!

**Comment by ZolotukhinM** [comment]
This looks like a bug in the current fuser. While we're looking at it, you could disable it by adding
```python
torch._C._jit_override_can_fuse_on_gpu(False)
```
to your script, and the issue should go away.

**Comment by ZolotukhinM** [comment]
I created a smaller reproducer that doesn't require CUDA. It uses some internal APIs, but it still can be used for further investigation:

```python
import torch
import torch._C._te as te

input_str = """
graph(%x : Half(3, 3, strides=[3, 1], requires_grad=0, device=cpu),
       %weight : Half(3, strides=[1], requires_grad=0, device=cpu),
       %bias : Half(3, strides=[1], requires_grad=0, device=cpu),
       %running_mean : Half(3, strides=[1], requires_grad=0, device=cpu),
       %running_var : Half(3, strides=[1], requires_grad=0, device=cpu)):
   %5 : bool = prim::Constant[value=1]()
   %6 : float = prim::Constant[value=0.001]()
   %7 : float = prim::Constant[value=0.10000000000000001]()
   %8 : bool = prim::Constant[value=0]()
   %input.1 : Half(3, 3, strides=[3, 1], requires_grad=0, device=cpu) = aten::batch_norm(%x, %weight, %bias, %running_mean, %running_var, %8, %7, %6, %5)
   %10 : Half(3, 3, strides=[3, 1], requires_grad=0, device=cpu) = aten::relu(%input.1)
   return (%10)
"""


class kernel_arena_scope(object):
    def __enter__(self):
        self.scope = torch._C._te.KernelScope()

    def __exit__(self, typ, val, traceback):
        self.scope = None

with kernel_arena_scope():
    graph = torch._C.parse_ir(input_str)
    print(graph)
    kernel = te.TensorExprKernel(graph) # Fails
    print(kernel.get_code_text("asm"))
```

**Comment by GrzegorzKarchNV** [comment]
> This looks like a bug in the current fuser. While we're looking at it, you could disable it by adding
> 
> ```python
> torch._C._jit_override_can_fuse_on_gpu(False)
> ```
> 
> to your script, and the issue should go away.

Thanks @ZolotukhinM. As far as I understand, `_jit_override_can_fuse_on_gpu(False)`  must be set at runtime, is that correct? If so, is there a way to set this flag as environment variable?

**Comment by ZolotukhinM** [comment]
That is correct, it should be set at runtime (but usually it can be done somewhere in the very beginning of the python script). AFAIR, it has no associated env variable, but I believe you should be able to easily use your own:
```python
if os.getenv('MY_VAR_TO_TURN_OFF_GPU_FUSION'):
  torch._C._jit_override_can_fuse_on_gpu(False)
```

**Comment by ZolotukhinM** [comment]
This should be fixed by #61569. If you use a nightly build or the next release, it should work. I'll close the issue for now, but please feel free to reopen if you need more help!

**Comment by GrzegorzKarchNV** [comment]
Thanks @ZolotukhinM !
